### PR TITLE
spec: overview of the p2p implementation in v0.34

### PR DESCRIPTION
**Pull request by cason**
Addresses #9089.

Description of the implementation of the p2p layer in v0.34. This is a work in progress.

### CODE REVIEW

File: spec/p2p/v0.34/switch.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+## AddReactor
+
+The reactor is associated to the set of channel ids it employs.
+Two reactors (in the same node) cannot share the same channel id.
+
+There is a call back to the reactor, in which the switch passes itself to the
+reactor.
+
+For some reason, the added reactor is returned by this method.
+
```
Comment (jmalicevic) [inline spec/p2p/v0.34/switch.md:30]: I think the return value is used only in tests. But there is a function in node.go to create a Custom reactor which calls `AddReactor`. But it is not used anywhere.

File: spec/p2p/v0.34/switch.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+effect and returns immediately.
+This is in fact a likely scenario, as the `reconnectToPeer` routine relies on
+this same `DialPeerWithAddress` method for dialing peers.
+
+### `DialPeersAsync` 
+
+This method receives a list of peer addresses (strings) and dials all of
+them in parallel. 
+
+> TODO: who invokes this? Apparently, it is invoked at startup to dial all
```
Comment (jmalicevic) [inline spec/p2p/v0.34/switch.md:156]: Yes these are the only places where this is called. The PEX reactor calls `DialPeerWithAddress`.

File: spec/p2p/v0.34/switch.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+This method broadcasts a message on a channel, by sending the message in
+parallel to all connected peers.
+
+The method spawns a thread for each connected peer, invoking the `Send` method
+provided by each `Peer` instance with the provided message and channel ID.
+The return value (a boolean) of these calls are redirected to a channel that is
+returned by the method.
+
+> TODO: does anyone use this method? Check.
```
Comment (jmalicevic) [inline spec/p2p/v0.34/switch.md:239]: I see it used in:
- consensus `broadcastNewRoundStepMessage`, ` broadcastNewValidBlockMessage` , 1broadcastHasVoteMessage`; 
- statesync (within `Sync`)

Comment (cason) [reply spec/p2p/v0.34/switch.md:239]: Ok, so it is in use.

File: spec/p2p/v0.34/switch.md (hunk 2 above)
Comment (cason) [reply spec/p2p/v0.34/switch.md:156]: So, we should document it as startup procedure.

File: spec/p2p/v0.34/switch.md (hunk 1 above)
Comment (cason) [reply spec/p2p/v0.34/switch.md:30]: This comment is not really relevant them.

File: spec/p2p/v0.34/pex-protocol.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+## Requesting Addresses
+
+A node requests peer addresses by sending a `PexRequest` message to a peer.
+
+For regular nodes, not operating in seed mode, a PEX request is sent when
+the node *needs* peers addresses, a condition checked:
+
+1. When an *outbound* peer is added, causing the node to request peer addresses
+   to the new peer
```
Comment (jmalicevic) [inline spec/p2p/v0.34/pex-protocol.md:27]: I would rephrase this as " request addresses from the new peer"

File: spec/p2p/v0.34/pex-protocol.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+As part of this procedure, the node selects a peer at random,
+from the set of connected peers retrieved from the switch,
+and sends a PEX request to the selected peer.
+
+Sending a PEX request to a peer is implemented by the `RequestAddrs` method of
+the PEX reactor.
+
+### Responses
+
+After sending a PEX request is sent to a peer, the node expect to receive,
```
Comment (jmalicevic) [inline spec/p2p/v0.34/pex-protocol.md:55]:
```suggestion
After a PEX request is sent to a peer, the node expects to receive,
```

File: spec/p2p/v0.34/pex-protocol.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+as a response, a `PexAddrs` message from the peer.
+This message encodes a list of peer addresses that are
+[added to address book](./addressbook.md#new-addresses),
```
Comment (jmalicevic) [inline spec/p2p/v0.34/pex-protocol.md:58]:
```suggestion
[added to the address book](./addressbook.md#new-addresses),
```

File: spec/p2p/v0.34/pex-protocol.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+having the peer from which the PEX response was received as their source.
+
+Received PEX responses are handled by the `ReceiveAddrs` method of the PEX reactor.
+In the case that of a PEX response received from a peer which is configured as
```
Comment (jmalicevic) [inline spec/p2p/v0.34/pex-protocol.md:62]:
```suggestion
In the case of a PEX response received from a peer which is configured as
```

File: spec/p2p/v0.34/pex-protocol.md [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+Sending a PEX response to a peer that has not requested peer addresses
+is also considered a misbehavior.
+So, if a PEX response is received from a peer that is not registered in
+the `requestsSent` set, a `ErrUnsolicitedList` error is produced.
+This leads the peer to be disconnected and marked as a bad peer.
+
+## Providing Addresses
+
+When a node receives a `PexRequest` message from a peer,
+it replies to peer with a `PexAddrs` message.
```
Comment (jmalicevic) [inline spec/p2p/v0.34/pex-protocol.md:86]:
```suggestion
it replies with a `PexAddrs` message.
```

File: spec/p2p/v0.34/pex-protocol.md [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+run the `ensurePeersRoutine`, as regular nodes,
+but run the `crawlPeersRoutine`, which is not run by regular nodes.
+
+The `crawlPeersRoutine` periodically, every 30 seconds (`crawlPeerPeriod`),
+starts a new peer discovery round.
+First, the seed node retrieves a random selection of peer addresses from its
+Address Book.
+This selection is produced in the same way as in the random selection of peer
+addresses that are [provided](#providing-addresses) to a requesting peer.
+Are removed from this selection peers that the seed node has crawled recently,
```
Comment (jmalicevic) [inline spec/p2p/v0.34/pex-protocol.md:139]:
```suggestion
From this selection peers that the seed node has crawled recently are removed,
```

File: spec/p2p/v0.34/pex-protocol.md [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+last than 2 minutes ago (`minTimeBetweenCrawls`).
```
Comment (jmalicevic) [inline spec/p2p/v0.34/pex-protocol.md:140]:
```suggestion
less than 2 minutes ago (`minTimeBetweenCrawls`).
```

File: spec/p2p/v0.34/pex-protocol.md [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+
+Moreover, the selection of peer addresses provided to inbound peers by a seed
+node, although still essentially random, has a [bias toward old
+addresses](./addressbook.md#random-selection-with-bias).
+The selection bias is defined by `biasToSelectNewPeers`, hard-coded to `30%`,
+meaning that `70%` of the peer addresses provided by a seed node are expected
+to be old addresses.
+Although this nomenclature is not clear, *old* addresses are the addresses that
+survived the most in the address book, that is, are addresses that the seed
+node believe that belong to *good* peers (more details [here](./addressbook.md#good-peers)).
```
Comment (jmalicevic) [inline spec/p2p/v0.34/pex-protocol.md:195]:
```suggestion
node believes belong to *good* peers (more details [here](./addressbook.md#good-peers)).
```

File: spec/p2p/v0.34/pex-protocol.md [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+If the same peer dials again to the seed node and requests peer addresses, the
+seed node will reply to this peer like it was the first time it has requested
+peer addresses.
+
+> This is more an implementation restriction than a desired behavior.
+> The `lastReceivedRequests` map stores the last time a PEX request was
+> received from a peer, and the entry relative to a peer is removed from this
+> map when the peer is disconnected.
+>
+> It is debatable whether this approach indeed prevent abuse against seed nodes.
```
Comment (jmalicevic) [inline spec/p2p/v0.34/pex-protocol.md:213]: While we should be concerned with abusing a seed node, we should also consider the scenario where in a network with few seed nodes a peer might be right to request addresses more then once. Maybe we can have an algorithm where we provide to the peer more then one address? Or each peer sends a desired number of addresses (this number can be derived from the currently connected peers and the desired number of peers)?

File: spec/p2p/v0.34/pex.md [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+## OnStart
+
+The `OnStart` method implements `BaseService` and starts the PEX reactor.
+
+The [address book](./addressbook.md), which is a `Service` is started.
+This loads the address book content from disk,
+and starts a routine that periodically persists the address book content to disk.
+
+The PEX reactor is configured with the address of a number of seed nodes,
```
Comment (jmalicevic) [inline spec/p2p/v0.34/pex.md:29]:
```suggestion
The PEX reactor is configured with the addresses of a number of seed nodes,
```

File: spec/p2p/v0.34/addressbook.md [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+Addresses that are not any longer banned, i.e., which banned period has expired,
+are added back to the address book as new addresses, while the corresponding
+node IDs are removed from the `badPeers` map.
+
+## Removing addresses
+
+The `RemoveAddress` method removes an address from the address book.
+
+It is invoked by the switch when it dials a peer or accepts a connection from a
+peer that end ups being the node itself (`IsSelf` error).
```
Comment (jmalicevic) [inline spec/p2p/v0.34/addressbook.md:188]: In which cases does a nodes own address end up being in the address book? In a lot of places I see a guard before adding a peer to the address book checking whether the address is ours. Unless this exists for the case we get this as persistent peers on startup. But even then I think there is a check before adding them to the address book.

File: spec/p2p/v0.34/addressbook.md [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+in the case that the node needs more peer addresses.
+
+The set of banned peer addresses is retrieved from the `badPeers` map.
+Addresses that are not any longer banned, i.e., which banned period has expired,
+are added back to the address book as new addresses, while the corresponding
+node IDs are removed from the `badPeers` map.
+
+## Removing addresses
+
+The `RemoveAddress` method removes an address from the address book.
```
Comment (jmalicevic) [inline spec/p2p/v0.34/addressbook.md:185]: The function is also called from the `MarkBad` function (the internal `removeAddress` function).

File: spec/p2p/v0.34/pex.md [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+
+If the peer is an **outbound peer**, i.e., if the node has dialed the peer,
+and the PEX protocol needs more addresses,
+the node [sends a PEX request](./pex-protocol.md#Requesting-Addresses) to the peer.
+The same is not done when inbound peers are added because they are deemed least
+trustworthy than outbound peers.
+
+## RemovePeer
+
+The `RemovePeer` method, from the `Reactor` interface,
```
Comment (jmalicevic) [inline spec/p2p/v0.34/pex.md:93]: This function does not actually remove the peer from the address book, right?

File: spec/p2p/v0.34/switch.md (hunk 1 above)
Comment (cason) [reply spec/p2p/v0.34/switch.md:30]: I've removed it.

File: spec/p2p/v0.34/switch.md (hunk 3 above)
Comment (cason) [reply spec/p2p/v0.34/switch.md:239]: Maintained as a TODO, but mentioning these uses.

File: spec/p2p/v0.34/pex-protocol.md (hunk 4 above)
Comment (cason) [reply spec/p2p/v0.34/pex-protocol.md:27]: Adopted.

File: spec/p2p/v0.34/pex-protocol.md (hunk 5 above)
Comment (cason) [reply spec/p2p/v0.34/pex-protocol.md:55]: Manually adopted.

File: spec/p2p/v0.34/pex-protocol.md (hunk 9 above)
Comment (cason) [reply spec/p2p/v0.34/pex-protocol.md:139]: I adopted the suggestion, but with a different format.

File: spec/p2p/v0.34/addressbook.md (hunk 14 above)
Comment (cason) [reply spec/p2p/v0.34/addressbook.md:188]: This is a good point. Although I believe this code is there for a practical reason :D

File: spec/p2p/v0.34/addressbook.md (hunk 15 above)
Comment (cason) [reply spec/p2p/v0.34/addressbook.md:185]: Good point, added to the document with a link to this section.

File: spec/p2p/v0.34/pex.md (hunk 16 above)
Comment (cason) [reply spec/p2p/v0.34/pex.md:93]: Nope. This is a method of the `Reactor` interface, which essentially means the peer was disconnected.

File: spec/p2p/v0.34/pex-protocol.md [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+## Requesting Addresses
+
+A node requests peer addresses by sending a `PexRequest` message to a peer.
+
+For regular nodes, not operating in seed mode, a PEX request is sent when
+the node *needs* peers addresses, a condition checked:
+
+1. When an *outbound* peer is added, causing the node to request addresses from
+   the new peer
+1. Periodically, by the `ensurePeersRoutine`, causing the node to request peer
```
Comment (JayT106) [inline spec/p2p/v0.34/pex-protocol.md:28]: `1.` -> `2.`

File: spec/p2p/v0.34/peer_manager.md [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+However, as observed in this [issue](https://github.com/tendermint/tendermint/issues/2093),
+it can take some time, up to `ensurePeersPeriod` or 30 seconds, from when the
+node receives new peer addresses and when it dials the received addresses.
+To avoid this delay, which can be particularly relevant when the node has no
+peers, a node immediately attempts to dial peer addresses when they are
+received from a peer that is locally configured as a seed node.
+
+> FIXME: The current logic was introduced in [#3762](https://github.com/tendermint/tendermint/pull/3762).
+> Although it fix the issue, the delay between receiving an address and dialing
+> the peer, it does not impose ant limit on how many addresses are dialed in this
```
Comment (JayT106) [inline spec/p2p/v0.34/peer_manager.md:85]: `ant` -> `and` ?

**Review by jmalicevic — APPROVED** [review]

File: spec/p2p/v0.34/pex-protocol.md (hunk 17 above)
Comment (jmalicevic) [reply spec/p2p/v0.34/pex-protocol.md:28]: Thank you for taking the time to review the spec. We have merged this PR because we are preparing a next one with new edits and will then open a PR against `main`. I will incorporated your suggestions there though.
